Add user interface for trades and trade offers

AEgan opened the pull request:
This PR adds controller actions and views for trading and offering trades in the system. It's a lot larger than I was expecting it to be.

Users are able to post a trade with one player, as you can do in the game. Users can browse through open trades separated by console and can view the details of the trade, including all offers for the trade. You can create an offer from the trade's show page, and chose an amount of coins (the in-game currency) as well as multiple players through a nested form. This will allow offers on the website, with a coin value and multiple players, to reflect how trades are done in the video game.

The user who posted the trade can then choose which offer they would like to accept, which would change the status of the trade and not allow other offers to come through or be modified.

6638c43 adds a validation to the offer model, which made me fix my testing context and modify tests.

It is [deployed on heroku](https://sleepy-bayou-2110.herokuapp.com) but needs a facelift. You can sign up or log in with username 'egan' and password 'secret'

There is an issue with the nested form for offers. Sometimes it works and adds one select box, other times it adds a seemingly random amount. I'm looking into that next.

@jeremy6d @jdlehman @nhessler

Commit:
adds trades controller, views, and routes. Adds table of trades to index page, tests index action and views

Commit:
add and test the new and create actions for the trades controller and their related views

Commit:
add information to the trades/show page, add tests for the show page and controller, add cancel button to cancel a trade

Commit:
add offer player model to replace the HABTM join table, and add tests for the offer player model

I decided to use the join table and the has_and_belongs_to_many relationship because I did not think
that I would need any additional checks or logic in the relationship between the two. After realizing
that in-game trade offers are limited to four players, and after having difficulty making a clean form
to allow for adding multiple players to an offer at once, I decided the best solution was to create
a model to represent the join table, place additional logic in that model, and replace the HABTM relationships
with the has_many through relationships. Unlike other tables with foreign keys, the offer_players
model does not check the reference on the offer because I am using a nested form, where the offer
will be created at the same time as the offer_players.

Commit:
add offers controller, new, create, show actions, and form

This adds an offers controller, nested resource routes, and views for
creating a trade offer. It also adds a table of offers on the trade show
page.

Commit:
add tests for offers_controller

Commit:
add edit and update actions to offers controller, add controller tests for the actions

Commit:
add offers show page and controller tests for show action

Commit:
add offers accept action and controller tests

Commit:
add ability to offer a trade and accept a trade through the ui

Commit:
add 2 validations around trading

This adds two validations around trading: users can only offer trades to the
same console, and users can send offers to open trades. This caused factories
being used in testing invalid, and I had to modify the contexts and all tests
using them.

Commit:
prepare for deploying changes to heroku

Commit:
creates authorization failure method to reuse code

Commit:
redirect user with prior offer to edit offer page if going to new

removing a TODO, when a user has already created a trade offer and goes to
create a new trade offer for the same trade, they will be redirected to the
edit page instead of being redirected home

jdlehman commented on app/controllers/offers_controller.rb:
```diff
@@ -0,0 +1,108 @@
+class OffersController < ApplicationController
+  before_action :set_trade, only: [:new, :show, :create, :edit, :update, :accept]
```
If you are using a `before_action` on all the actions in a controller I believe you can leave off the `only`

``` ruby
before_action :set_trade
```

Commit:
increase readability of before_action calls and refactor code

remove only: on before action called for everything, fix typo,
remove unnecessary logged_in? call, refactors an authorization check
I was making multiple times, replace only: with except: where it made
sense, and replace #nil? with #blank?

AEgan commented:
I just realized I misspelled `create_offers` doing a find/replace all. I'll push that with other fixes

jdlehman commented on app/controllers/offers_controller.rb:
```diff
@@ -0,0 +1,108 @@
+
+  def check_user_on_same_console
+    user = @trade.user
+    if !logged_in? || current_user.console != user.console
```
Is there a reason you need to check `!logged_in?` here?  Everywhere else you are working with the `current_user` object without this check.

If you need to ensure they are logged in, I would probably make it a separate check in the `before_action`s

jdlehman commented on app/controllers/trades_controller.rb:
```diff
@@ -0,0 +1,55 @@
+  end
+
+  def cancel
+    if !logged_in? || current_user.id != @trade.user_id || @trade.status != Trade::STATUSES['open']
```
You are using `current_user.id != @trade.user_id` in several places. It might warrant pulling  some of these reused equality checks into their own methods so you can rename them in a way that makes it easier to read what the code is checking as well as get the benefit of reuse too.

jdlehman commented on app/helpers/trades_helper.rb:
```diff
@@ -0,0 +1,5 @@
+module TradesHelper
+  def user_can_offer_trade?
+    logged_in? && current_user.id != @trade.user_id && @trade.status == Trade::STATUSES['open'] && current_user.console == @user.console
```
Wow this is quite a check, great job pulling it into its own easy to read method.

Like I mentioned in another comment some of these might also be worth pulling into their own methods, but obviously you could go overboard on that, which wouldn't help the code either.  Just pointing it out as an area to evaluate, not necessarily something that needs to be done.  Part of it is preference

jdlehman commented on app/views/offers/show.html.erb:
```diff
@@ -0,0 +1,31 @@
+  There are no players associated with this trade
+<% else %>
+  <h2>Players</h2>
+  <table>
```
tables definitely get the job done, but I have found that using divs makes it a lot easier to tweak the formatting or completely redesign something in the future without having to throw everything away.

This also will allow you to write modular CSS  such that your classes help provide context to other people reading your code.  If you are interested, here is an introductory [article](http://inlehmansterms.net/2014/08/25/modular-css-with-suitcss/) on how we write modular Sass at CustomInk with SUIT CSS

jdlehman commented on app/views/trades/show.html.erb:
```diff
@@ -0,0 +1,65 @@
+      <tbody>
+        <% @offers.each do |offer| %>
+          <tr>
+            <td><%= offer.user.username %></td>
```
These views are probably not quite large enough to warrant it, but might be worth thinking about splitting some things into partials to logically split up parts of the view on a page.

jdlehman commented:
I left some comments and suggestions inline (some more nitpicky than others).  Also, this is a really cool app idea, I am impressed!

AEgan commented:
Thanks for the feedback! I'll make some fixes based on your comments, you definitely pointed out some things I overlooked or didn't know about.

jeremy6d commented on app/controllers/application_controller.rb:
```diff
@@ -14,7 +14,7 @@ def sign_in(user)
 
   def current_user
-    @current_user ||= User.find(session[:user_id]) if session[:user_id]
+    @current_user ||= User.find_by_id(session[:user_id]) if session[:user_id]
```
What was the reasoning behind this change?  Aren't `#find` and `#find_by_id` equivalent?

AEgan replied:
I made this change mostly because I've been dropping and recreating the database a lot when I modify the seeds file when I add tables to the database. If I'm logged in when I drop the database, the next time I open up the app I'd get an ActiveRecord::RecordNotFound exception, which is raised when `#find` doesn't find the record. `#find_by_id` will return nil, and will set `current_user` to nil. It just makes developing a bit easier when I'm playing with populating the database.

jeremy6d replied:
Great explanation!

jeremy6d commented on app/controllers/application_controller.rb:
```diff
@@ -30,4 +30,15 @@ def logged_in?
+    end
+  end
+
+  def authorization_failure(msg = "You are not authorized to perform this action.")
```
For what it's worth, I think the way you did this is pretty slick.  I've done authorization failure handlers before but never thought of providing an overridable message to the user.

jdlehman replied:
:+1: I concur

jeremy6d commented on app/controllers/offers_controller.rb:
```diff
@@ -0,0 +1,108 @@
+class OffersController < ApplicationController
+  before_action :set_trade, only: [:new, :show, :create, :edit, :update, :accept]
+  before_action :set_offer, only: [:show, :edit, :update, :accept]
```
Similarly to what Jonathan said, if you want the method to run on _most_ actions, you can do: '

`before_action :set_offer, except: [:new, :create]`  

Rails can be both a blessing and a curse in the diversity of constructs it offers; it's always a human judgment call about which expresses the intent better. :smile:

jeremy6d commented on app/controllers/offers_controller.rb:
```diff
@@ -0,0 +1,108 @@
+  end
+
+  def create
+    @offer = Offer.new(offer_params.merge({trade_id: @trade.id, user_id: current_user.id}))
```
Great use of `#merge` here!

jeremy6d commented on app/views/users/show.html.erb:
```diff
@@ -3,7 +3,7 @@
 <p>Email: <%= @user.email %></p>
 <p>Console: <%= @user.console_name %></p>
 <p>Team Name: <%= @user.team_name %></p>
-<p>Reputation: <%= @user.reputation.nil? ? "No trades yet" : @user.reputation %></p>
```
It might be better practice here to use `#blank?` instead of `#nil?` since the former checks against empty strings as well.

jeremy6d commented on config/routes.rb:
```diff
@@ -4,6 +4,11 @@
   post 'sessions' => 'sessions#create'
   resources :users, only: [:show, :new, :edit, :update, :create]
   resources :teams, only: [:show, :index]
+  resources :trades, only: [:index, :show, :new, :edit, :update, :create] do
```
Another example of where a `except: [:delete]` might be more expressive.

AEgan commented:
Thanks for the feedback!
